File: gptcomet/llms/cohere.py
from typing import Any, Optional

from .base import BaseLLM


class CohereLLM(BaseLLM):
    """Cohere LLM provider implementation."""

    def __init__(self, config: dict[str, Any]):
        super().__init__(config)

        self.api_base = config.get("api_base") or "https://api.cohere.ai/v1"
        self.model = config.get("model") or "command-r"
        self.answer_path = config.get("answer_path") or "text"
        self.completion_path = config.get("completion_path") or "/chat"

    def build_headers(self):
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def format_messages(
        self, message: str, history: Optional[list[dict[str, str]]] = None
    ) -> dict[str, Any]:
        """Format messages for Cohere API."""
        chat_history = []

        if history:
            for msg in history:
                chat_history.append(
                    {
                        "role": "CHATBOT" if msg["role"] == "assistant" else "USER",
                        "message": msg["content"],
                    }
                )

        payload = {
            "message": message,
            "model": self.model,
            "chat_history": chat_history,
        }
        if self.max_tokens is not None:
            payload["max_tokens"] = int(self.max_tokens)
        if self.temperature is not None:
            payload["temperature"] = float(self.temperature)
        if self.top_p is not None:
            payload["top_p"] = float(self.top_p)
        if self.frequency_penalty is not None:
            payload["frequency_penalty"] = float(self.frequency_penalty)
        if self.presence_penalty is not None:
            payload["presence_penalty"] = float(self.presence_penalty)

        return payload

    def get_usage(self, data: dict[str, Any]) -> Optional[str]:
        usage = data.get("meta", {}).get("billed_units")
        if not usage:
            return None
        else:
            return f"Token usage> input tokens: {usage.get('input_tokens')}, output tokens: {usage.get('output_tokens')}"

    @classmethod
    def get_required_config(cls) -> dict[str, tuple[str, str]]:
        return {
            "api_base": ("https://api.cohere.ai/v1", "Enter Cohere API Base URL"),
            "model": ("command-r", "Enter model name"),
            "api_key": ("", "Enter API key"),
            "max_tokens": ("1024", "Enter max tokens"),
        }

File: internal/llm/cohere.go
package llm

import (
	"fmt"

	"github.com/belingud/go-gptcomet/pkg/config"
	"github.com/belingud/go-gptcomet/pkg/types"
	"github.com/tidwall/gjson"
)

// CohereLLM implements the LLM interface for Cohere
type CohereLLM struct {
	*OpenAILLM
}

// NewCohereLLM creates a new CohereLLM
func NewCohereLLM(config *types.ClientConfig) *CohereLLM {

	if config.APIBase == "" {
		config.APIBase = "https://api.cohere.com/v2"
	}
	if config.Model == "" {
		config.Model = "command-r-plus"
	}

	return &CohereLLM{
		OpenAILLM: NewOpenAILLM(config),
	}
}

func (c *CohereLLM) Name() string {
	return "cohere"
}

// GetRequiredConfig returns provider-specific configuration requirements
func (c *CohereLLM) GetRequiredConfig() map[string]config.ConfigRequirement {
	return map[string]config.ConfigRequirement{
		"api_base": {
			DefaultValue:  "https://api.cohere.com/v2",
			PromptMessage: "Enter Cohere API base",
		},
		"api_key": {
			DefaultValue:  "",
			PromptMessage: "Enter API key",
		},
		"model": {
			DefaultValue:  "command-r-plus",
			PromptMessage: "Enter model name",
		},
		"max_tokens": {
			DefaultValue:  "1024",
			PromptMessage: "Enter max tokens",
		},
	}
}

// FormatMessages formats messages for Cohere API
func (c *CohereLLM) FormatMessages(message string, history []types.Message) (interface{}, error) {

	messages := []map[string]string{
		{
			"role":    "user",
			"content": message,
		},
	}

	payload := map[string]interface{}{
		"messages":    messages,
		"model":       c.Config.Model,
		"stream":      false,
		"max_tokens":  c.Config.MaxTokens,
		"temperature": c.Config.Temperature,
	}

	return payload, nil
}

// GetUsage returns usage information for the provider
func (c *CohereLLM) GetUsage(response []byte) (string, error) {
	usage := gjson.GetBytes(response, "usage")
	if !usage.Exists() {
		return "", nil
	}

	return fmt.Sprintf(
		"Token usage> input: %d, output: %d",
		usage.Get("input_tokens").Int(),
		usage.Get("output_tokens").Int(),
	), nil
}
